opm/autodiff/AquiferFetkovich.hpp

```diff
@@ -0,0 +1,387 @@
+        {
+          const auto& fs = intQuants.fluidState();
+          pressure_water.at(idx) = fs.pressure(waterPhaseIdx).value();
+        }
+
+        inline void updateCellDensity(const int idx, const IntensiveQuantities& intQuants)
+        {
+          const auto& fs = intQuants.fluidState();
+          rhow_.at(idx) = fs.density(waterPhaseIdx);
+        }
+
+        inline Scalar dpai(int idx)
+        {
+          Scalar dp = aquifer_pressure_ + rhow_.at(idx).value()*gravity_()*(cell_depth_.at(idx) - aqufetp_data_.d0) - pressure_current_.at(idx).value() ;
+          return dp;
+        }
+        // This function implements Eq 5.12 of the EclipseTechnicalDescription
+        inline Scalar aquiferPressure()
+        {
+          Scalar Flux = W_flux_.value();
+          Scalar pa_ = pa0_ - Flux / ( aqufetp_data_.C_t * aqufetp_data_.V0 );
+          return pa_;
+        }
+        // This function implements Eq 5.14 of the EclipseTechnicalDescription
+        inline void calculateInflowRate(int idx, const Simulator& simulator)
+        {
+          Tc_ = ( aqufetp_data_.C_t * aqufetp_data_.V0 ) / aqufetp_data_.J ;
+          Scalar td_Tc_ = simulator.timeStepSize() / Tc_ ;
+          Scalar exp_ = (1 - exp(-td_Tc_)) / td_Tc_;
+          Qai_.at(idx) = alphai_.at(idx) * aqufetp_data_.J * dpai(idx) * exp_;
+        }
+
+        template<class faceCellType, class ugridType>
+        inline const double getFaceArea(const faceCellType& faceCells, const ugridType& ugrid,
+                                        const int faceIdx, const int idx,
+                                        const Aquancon::AquanconOutput& connection) const
+          {
+            // Check now if the face is outside of the reservoir, or if it adjoins an inactive cell
+            // Do not make the connection if the product of the two cellIdx > 0. This is because the
+            // face is within the reservoir/not connected to boundary. (We still have yet to check for inactive cell adjoining)
+            double faceArea = 0.;
+            const auto cellNeighbour0 = faceCells(faceIdx,0);
+            const auto cellNeighbour1 = faceCells(faceIdx,1);
+            const auto defaultFaceArea = Opm::UgGridHelpers::faceArea(ugrid, faceIdx);
+            const auto calculatedFaceArea = (!connection.influx_coeff.at(idx))?
+            defaultFaceArea :
+            *(connection.influx_coeff.at(idx));
+            faceArea = (cellNeighbour0 * cellNeighbour1 > 0)? 0. : calculatedFaceArea;
+            if (cellNeighbour1 == 0){
+              faceArea = (cellNeighbour0 < 0)? faceArea : 0.;
+            }
+            else if (cellNeighbour0 == 0){
+              faceArea = (cellNeighbour1 < 0)? faceArea : 0.;
+            }
+            return faceArea;
+          }
+
+          // This function is used to initialize and calculate the alpha_i for each grid connection to the aquifer
+          inline void initializeConnections(const Aquancon::AquanconOutput& connection)

```

Comment:
How different is this function from the version in `AquiferCarterTracy`. If they are the same, maybe with some refactoring, you can also put this function to the Helper file to reduce some duplication of code. Alternatively, you can put it in a base class for `AquiferFetkovich` and `AquiferCarterTracy`.